src/operators/mpas_vector_reconstruction.F

```diff
@@ -233,64 +244,108 @@ subroutine mpas_reconstruct_2d(meshPool, u, uReconstructX, uReconstructY, uRecon
     call mpas_pool_get_array(meshPool, 'edgesOnCell', edgesOnCell)
 
     if ( includeHalosLocal ) then
-       call mpas_pool_get_dimension(meshPool, 'nCells', nCells)
+       call mpas_pool_get_dimension(meshPool, 'nCells', nCells_ptr)
     else
-       call mpas_pool_get_dimension(meshPool, 'nCellsSolve', nCells)
+       call mpas_pool_get_dimension(meshPool, 'nCellsSolve', nCells_ptr)
     end if
+    call mpas_pool_get_dimension(meshPool, 'nVertLevels', nVertLevels_ptr)
+    ! Dereference pointers so that OpenACC copies them correctly
+    nCells = nCells_ptr
+    nVertLevels = nVertLevels_ptr
 
     call mpas_pool_get_array(meshPool, 'latCell', latCell)
     call mpas_pool_get_array(meshPool, 'lonCell', lonCell)
 
     call mpas_pool_get_config(meshPool, 'on_a_sphere', on_a_sphere)
 
+    MPAS_ACC_TIMER_START('mpas_reconstruct_2d [ACC_data_xfer]')
+    ! Only use sections needed, nCells may be all cells or only non-halo cells
+    !$acc enter data copyin(coeffs_reconstruct(:,:,1:nCells),nEdgesOnCell(1:nCells), &
+    !$acc                   edgesOnCell(:,1:nCells),latCell(1:nCells),lonCell(1:nCells))
```

Comment:
Copying in subsets of array isn't something that we've done in other places, and although this seems solid, I'm wondering whether there might be any ways in which this could lead to problems? For example, if we were to only copy in the solve cells for `latCell` in `mpas_atm_dynamics_init` but `nCells` here were defined as all cells (owned plus halo), will the on-device arrays be properly re-alloc'd to accommodate the full range of cells?

Reply:
There's no problems I can tell from running a test with only copying in `latCell(1:nCellSolve)` in the dynamics init routine. I think the OpenACC runtime can tell what's going on here - especially since the start of the memory is the same. Using more complicated array slices could get us in to trouble in the future.